onnxruntime/core/graph/function_utils.cc

```diff
@@ -0,0 +1,672 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/graph/function_utils.h"
+#include "core/common/inlined_containers.h"
+#include "core/framework/tensorprotoutils.h"
+#include "onnx/shape_inference/implementation.h"
+#include "core/graph/function_impl.h"
+#include "core/graph/model_load_utils.h"
+
+namespace ONNX_NAMESPACE {
+// Infer shape for functions. This also supports
+// nested model local functions.
+// TODO: Add this to onnx instead of adding it here.
+void InferShapeForFunctionNode(
+    InferenceContext& ctx,
+    const FunctionProto& func_proto,
+    const std::unordered_map<std::string, int>& func_opset_imports,
+    const ShapeInferenceOptions& options,
+    const ISchemaRegistry* schema_registry,
+    const std::unordered_map<std::string, const FunctionProto*>& in_model_functions,
+    std::function<std::string(const std::string& function_domain, const std::string& function_name)> get_func_id) {
+  GraphProto g;
+  // Get a temporary tensor-shape map
+  const auto num_func_inputs = func_proto.input_size();
+  std::unordered_map<std::string, TypeProto*> value_types_by_name;
+  std::vector<TypeProto> types_cache(func_proto.input_size());
```

Comment:
>(func_proto.input_size()) [](http://example.com/codeflow?start=36&length=25)

This defult constructs TypeProto sand then it throws it away. Please, change to reserve(), push_back().